front/lib/api/assistant/agent_usage.ts

```diff
@@ -22,261 +13,134 @@ import { redisClient } from "@app/lib/redis";
-                workspaceId: owner.id,
-              },
-            },
-          ],
-        },
-      ],
-    });
-    for (const mention of mentions) {
-      // No need to promise.all() here, as one Redis connection can only execute one command
-      // at a time.
-      if (mention.message?.userMessage) {
-        await signalInRedis({
-          agentConfigurationId,
+    // agent mention count doesn't exist
+    if (agentMessageCountTTL === -2) {
+      const [agentMessageCounts, userCounts] = await Promise.all([
+        agentMentionsCount(owner.id),
+        agentMentionsUserCount(owner.id),
+      ]);
+      await storeCountsInRedis(
+        workspaceId,
+        agentMessageCounts,
+        userCounts,
+        redis
+      );
+      // agent mention count is stale
+    } else if (agentMessageCountTTL < popularityComputationTimeframeSec) {
+      void (async () => {
+        const [agentMessageCounts, userCounts] = await Promise.all([
+          agentMentionsCount(owner.id),
+          agentMentionsUserCount(owner.id),
+        ]);
+        void storeCountsInRedis(
           workspaceId,
-          userId:
-            mention.message.userMessage.userId?.toString() ||
-            mention.message.userMessage.userContextEmail ||
-            mention.message.userMessage.userContextUsername,
-          timestamp: mention.createdAt.getTime(),
-          redis,
-        });
-      }
+          agentMessageCounts,
+          userCounts,
+          redis
+        );
+      })();
     }
-  }
-}
-
-// We don't need real-time accuracy on the number of members
-// that added the agent in their list.
-// Best-effort, we cache the result for a day.
-async function getAgentInListCount(
-  auth: Authenticator,
-  redis: Awaited<ReturnType<typeof redisClient>>,
-  workspaceId: string,
-  agentConfiguration: LightAgentConfigurationType
-) {
-  const { agentInListCountKey } = _getKeys({
-    agentConfigurationId: agentConfiguration.sId,
-    workspaceId,
-  });
-
-  const cachedCount = await redis.get(agentInListCountKey);
-  if (cachedCount !== null) {
-    return parseInt(cachedCount, 10);
-  }
-
-  const calculateAndCacheCount = async (countPromise: Promise<number>) => {
-    const count = await countPromise;
-    await redis.set(agentInListCountKey, count.toString(), {
-      EX: 86400, // Cache for 1 day.
-    });
-    return count;
-  };
 
-  // Determine the count based on the scope and cache the result.
-  switch (agentConfiguration.scope) {
-    case "published":
-      return calculateAndCacheCount(
-        getUsersWithAgentInListCount(auth, agentConfiguration.sId)
-      );
-    case "workspace":
-      return calculateAndCacheCount(
-        getMembersCount(auth, { activeOnly: true })
-      );
-    case "global":
-    case "private":
-      return 0;
-    default:
-      assertNever(agentConfiguration.scope);
+    // Retrieve and parse agents usage
+    const agentsUsage = await redis.hGetAll(agentMessageCountKey);
```

Comment:
if we GET outside the ttl (and the above query hasn't finished), will we still get the old values?

Reply:
There's two scenarios:
1. The values are stale (after 2h) but we are not outside the TTL (4h): we return the old value and trigger an update in the background
2. We are outside the TTL: we wait for the value to be computed

Reply:
LGTM